### ISSUE TITLE
Importing gunicorn.config takes a lot of time

### ISSUE THREAD

**Issue by isidentical** [benoitc/gunicorn#2613]
The import-time of certain packages are quite important for CLI applications, since even `0.1` seconds on each invocation adds up to something. While I was trying to figure out why importing `aiohttp` takes so long I noticed that the majority of the time spent on importing `gunicorn.config` which imported `gunicorn.utils` which then resulted with the import of `pkg_resources` (which is notoriously known for bad import times when the environment have too many packages);
```
 $ python -Ximporttime -c 'import gunicorn.config' 2>&1 | tail
import time:       424 |        424 |       sysconfig
import time:    215305 |     253223 |     pkg_resources
import time:       173 |        173 |     gunicorn.errors
import time:       113 |        113 |     gunicorn.workers
import time:       242 |        242 |     setproctitle
import time:       301 |     265796 |   gunicorn.util
import time:       229 |        229 |       inotify
import time:        18 |        246 |     inotify.adapters
import time:       244 |        490 |   gunicorn.reloader
import time:      3049 |     293565 | gunicorn.config
```

There is a [very long issue](https://github.com/pypa/setuptools/issues/510) (100+ comments) on the setuptools tracker about why this is slow but the important thing is that, gunicorn doesn't necessarily need to import it every time. There is only a single function which uses this to select entrypoints, so it might be worth considering to inline this import;
https://github.com/benoitc/gunicorn/blob/1299ea9e967a61ae2edebe191082fd169b864c64/gunicorn/util.py#L57-L104

```
 $ python -Ximporttime -c 'import gunicorn.config' 2>&1 | tail
import time:        38 |         38 |       atexit
import time:      1853 |       2203 |     logging
import time:       145 |        145 |     gunicorn.errors
import time:       108 |        108 |     gunicorn.workers
import time:       239 |        239 |     setproctitle
import time:       315 |      12995 |   gunicorn.util
import time:       235 |        235 |       inotify
import time:        18 |        252 |     inotify.adapters
import time:       213 |        465 |   gunicorn.reloader
import time:      2981 |      39848 | gunicorn.config
```

With the inline, on my environment (I have too many packages, but still) it went from `0.3` seconds to `0.039848` seconds. It also speeds up `aiohttp` considerably. 

master;
```
 $ python -Ximporttime -c 'import aiohttp' 2>&1 | tail
import time:       314 |     397838 | aiohttp
```

my branch;
```
 $ python -Ximporttime -c 'import aiohttp' 2>&1 | tail
import time:       292 |     144031 | aiohttp
```

**Comment by jamadden** [comment]
Can you provide more information about your use-case? Caring about import time seems atypical, because gunicon isn't meant to be a transient CLI program but rather a long-lived server.

**Comment by isidentical** [comment]
As I have stated, this actually traces back to a dependency (aiohttp) which is commonly used in CLI programs
https://github.com/aio-libs/aiohttp/blob/60f57272ab1ac230b2e5a3c8eca9f3ccf9c24252/aiohttp/worker.py#L11-L12

**Comment by jamadden** [comment]
That just makes it their problem then. Why is `aiohttp` importing a server package (workers) when not needed? I suggest solving the problem at a higher level.

**Comment by isidentical** [comment]
Is there anything I am missing about why not to inline a trivial import to call sites? It seems like it won't cause any regressions, and it will actually help everyone.

**Comment by jamadden** [comment]
I'm trying to help you find a better solution. This project tends to have a very long release cycle, so even a trivial change is unlikely to be seen on PyPI for quite some time.

Plus, this change won't help everyone, it will help those that are using gunicorn in very unusual use cases (importing it in short-lived CLI programs).

**Comment by isidentical** [comment]
> I'm trying to help you find a better solution. This project tends to have a very long release cycle, so even a trivial change is unlikely to be seen on PyPI for quite some time.

I see. Thanks for responding! Will try to figure out this on the higher level.

**Comment by m-novikov** [comment]
I also have the same problem, for me its noticeable during running the test runtime. with the same setup aiohttp import gunicorn. I don't think this is upstream problem. Worker is needed to setup the app, but the guniron import time is problematic.

**Comment by m-novikov** [comment]
> Plus, this change won't help everyone, it will help those that are using gunicorn in very unusual use cases (importing it in short-lived CLI programs). 

I disagree. It can significantly improve startup time of certain apps. Depending on codebase structure `import pkg_resources` step could be measured in seconds.

**Comment by hauntsaninja** [comment]
Looks like this is now mostly fixed, in that `gunicorn.util` no longer imports pkg_resources. Also aiohttp added a workaround for this in https://github.com/aio-libs/aiohttp/pull/6591

**Comment by benoitc** [comment]
no activity since awhile. closing feel free to create a new ticket if needed.

**Comment by cwegener** [comment]
Just confirming that this is fixed in gunicorn 21.0.0 as pointed out by https://github.com/benoitc/gunicorn/pull/2539#issuecomment-1712998953
